### PR TITLE
Introduce notion of `malloc` and `free`

### PR DESCRIPTION
**Pull request by chadhietala**
### Update

So there are some things that I would like people to be made of aware of here:

- This compiles each template into it's own `MemorySlab`
- Execution cannot "long jump" into another `MemorySlab`
- Inline blocks are a bit weird and [inherit the caller's `MemorySlab`](https://github.com/glimmerjs/glimmer-vm/pull/490/files#diff-2759ec8c9fd50aa5e6b288bb0053c1e5R67)
- Arrays on `Constants` are now lazily allocated
- `MemorySlab`s are pooled and recycled

--------------------

This introduces the notion of a [malloc](http://moss.cs.iit.edu/cs351/slides/slides-malloc.pdf) 👋 👋 👋. Effectively we want to have slabs of memory for the updating VM that we can free when a `Test` transitions or `Assert` fails. Today the updating VM does not use the stack VM because there is no notion of freeing the JIT'd program. Instead the updating VM creates LinkedLists of instructions that can just be GCd when they are no longer needed.

### Changes To Program and Constants
Prior to this commit we would have one slab of memory for the append [VM's program](https://github.com/glimmerjs/glimmer-vm/blob/master/packages/@glimmer/runtime/lib/environment.ts#L264-L303) that would dynamically grow and hold operations and pointers into a [heap](https://github.com/glimmerjs/glimmer-vm/blob/master/packages/@glimmer/runtime/lib/environment/constants.ts#L15-L101). This PR changes how these constructs are allocated. Instead of creating a single instance of the Program and Constants then attaching them to the environment we create some indirection so that we can pool them. The `Memory` class groups these two into a `MemorySlab` which the rest of the system writes to and what is freed. This allows us to free both the compiled program and any constants associated with the program. `Memory` is simply a manager for the `nextFree` pointer and the array of these `MemorySlabs`.

### Other Changes
- We now allocate the arrays on Constants lazily
- Constants is recyclable

### CODE REVIEW

File: tsconfig.json [hunk 1]
```diff
@@ -21,24 +21,19 @@
...
     "noImplicitReturns": true,
     "noImplicitThis": true,
-
+    "listFiles":false,
     "newLine": "LF",
     "noEmit": true,
-
     "paths": {
-      "*": ["*"]
+      "@glimmer/wire-format": [
```
Comment (rwjblue) [inline tsconfig.json]: I'm confused by these changes, what do they do?

Comment (chadhietala) [inline tsconfig.json]: Ignore these for now, going to remove before I'm done.

File: packages/@glimmer/runtime/lib/syntax/functions.ts [hunk 2]
```diff
@@ -925,11 +925,18 @@ export function compileStatement(statement: WireFormat.Statement, builder: Opcod
...
   finalize(): number;
 } {
-  let b = new OpcodeBuilder(env, meta);
+
+  let memory: MemorySlab;
+  if (parentSlab !== undefined) {
+    memory = env.memory.slab(parentSlab);
+  }
+
+  let b = new OpcodeBuilder(env, meta, env.memory);
```
Comment (krisselden) [inline packages/@glimmer/runtime/lib/syntax/functions.ts]: didn't you mean to pass in the memory here?

Comment (chadhietala) [inline packages/@glimmer/runtime/lib/syntax/functions.ts]: You're right

**Comment by chadhietala** [comment]
So there are some things that I would like people to be made of aware of here:

- This compiles each template into it's own `MemorySlab`
- Execution cannot "long jump" into another `MemorySlab`
- Inline blocks are a bit weird and [inherit the caller's `MemorySlab`](https://github.com/glimmerjs/glimmer-vm/pull/490/files#diff-2759ec8c9fd50aa5e6b288bb0053c1e5R67)
- Arrays on `Constants` are now lazily allocated
- `MemorySlab`s are pooled and recycled

**Comment by chadhietala** [comment]
Using glimmer-bench, these are the results:

### Established Markets Test
```
Insignificant U Value For compile: [417.5,482.5]

compile baseline vs. experiment

baseline      ·•[[======|===]-]   •·
experiment    ·•[---[===|===]--]             •·


Insignificant U Value For run: [528,372]

run baseline vs. experiment

baseline      ·•       [-----[===|===]-]  •·
experiment           ·• [--[===|==]----]     •·


Insignificant U Value For js: [408,492]

js baseline vs. experiment

baseline        ·•[-[=========|====]----] •·
experiment    ·•   [--[========|===]----]    •·


Insignificant U Value For gc: [447,453]

gc baseline vs. experiment

baseline      ·•   [---[===|====]----]      •·
experiment         ·•[--[==|===]---------]   •·


Insignificant U Value For callFunction: [336,564]

callFunction baseline vs. experiment

baseline      ·•[-[======|===]--------]      •·
experiment       ·•[--[=====|===]---] •·


Insignificant U Value For parseOnBackground: [515,385]

parseOnBackground baseline vs. experiment

baseline      ·•[[====================|==]-] •·
experiment    ·•[[===================|=]-]•·


Insignificant U Value For duration: [414,486]

duration baseline vs. experiment

baseline      ·•  [--[====|=====]----]      •·
experiment    ·•  [----[===|===]---]         •·


Insignificant U Value For usedHeapSizeBefore: [450,450]

usedHeapSizeBefore baseline vs. experiment

baseline·•[[|]]•·
experiment·•[[|]]•·


Insignificant U Value For usedHeapSizeAfter: [450,450]

usedHeapSizeAfter baseline vs. experiment

baseline·•[[|]]•·
experiment·•[[|]]•·


Self
Insignificant U Value For load: [523,377]

load baseline vs. experiment

baseline      ·•   [---[====|==]----]        •·
experiment        ·•[--[===|==]----]    •·


Cumulative
Insignificant U Value For load: [523,377]

load baseline vs. experiment

baseline      ·•   [---[====|==]----]        •·
experiment        ·•[--[===|==]----]    •·


Self
Insignificant U Value For render: [367.5,532.5]

render baseline vs. experiment

baseline      ·• [--[====|======]-------]    •·
experiment        ·•[---[=====|==]-----] •·


Cumulative
Insignificant U Value For render: [414,486]

render baseline vs. experiment

baseline      ·•  [--[====|=====]----]      •·
experiment    ·•  [----[===|===]---]         •·

```

### Emerging Markets Test
```
Insignificant U Value For compile: [476,424]

compile baseline vs. experiment

baseline      ·•[[==========|=====]---]  •·
experiment    ·•[[=========|========]--]     •·


Insignificant U Value For run: [471,429]

run baseline vs. experiment

baseline          ·•[[=|==]-----------]      •·
experiment    ·•  [-[==|=====]---]  •·


Insignificant U Value For js: [493,407]

js baseline vs. experiment

baseline       ·•[[=======|==]---]           •·
experiment    ·•[[=====|=====]--]      •·


Insignificant U Value For gc: [477,423]

gc baseline vs. experiment

baseline       ·• [-----[======|===]------]  •·
experiment    ·•    [--[======|====]-----] •·


Insignificant U Value For callFunction: [499,401]

callFunction baseline vs. experiment

baseline      ·•[[===|====]---]              •·
experiment    ·•[-[=|====]-]                •·


Insignificant U Value For parseOnBackground: [438,462]

parseOnBackground baseline vs. experiment

baseline      ·•[[===================|=]-]   •·
experiment    ·•[[========|============]] •·


Insignificant U Value For duration: [461,439]

duration baseline vs. experiment

baseline      ·•[-[==|=]--------]            •·
experiment    ·•[-[==|=]-]                 •·


Insignificant U Value For usedHeapSizeBefore: [450,450]

usedHeapSizeBefore baseline vs. experiment

baseline·•[[|]]•·
experiment·•[[|]]•·


Insignificant U Value For usedHeapSizeAfter: [450,450]

usedHeapSizeAfter baseline vs. experiment

baseline·•[[|]]•·
experiment·•[[|]]•·


Self
Insignificant U Value For load: [451,449]

load baseline vs. experiment

baseline           ·•[[==|=]-----------]     •·
experiment    ·•  [---[==|==]--]    •·


Cumulative
Insignificant U Value For load: [451,449]

load baseline vs. experiment

baseline           ·•[[==|=]-----------]     •·
experiment    ·•  [---[==|==]--]    •·


Self
Insignificant U Value For render: [473,427]

render baseline vs. experiment

baseline      ·•[[===|===]--]          •·
experiment    ·• [-[|==]-]                   •·


Cumulative
Insignificant U Value For render: [461,439]

render baseline vs. experiment

baseline      ·•[-[==|=]--------]            •·
experiment    ·•[-[==|=]-]                 •·
```

**Comment by chadhietala** [comment]
I'm going to close this for now but going to keep the branch around.

**Pull request closed without merging**